realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt

```diff
@@ -389,66 +363,151 @@ class MongoCollectionTest {
+//            assertEquals(3, count().blockingGetResult())
+//
+//            val options6 = FindOneAndModifyOptions()
+//                    .projection(sampleProject)
+//                    .sort(Document("num", -1))
+//            val result6 = findOneAndUpdate(
+//                    Document(),
+//                    sampleUpdate,
+//                    options6
+//            ).blockingGetResult()
+//            assertEquals(Document("hello", "world3"), result6!!.withoutId())
+//            assertEquals(3, count().blockingGetResult())
 
             // Test proper failure
-//            try {
-//                Tasks.await(coll.findOneAndUpdate(Document(), Document("\$who", 1)))
-//                fail()
-//            } catch (ex: ExecutionException) {
-//                assertTrue(ex.cause is StitchServiceException)
-//                val svcEx = ex.cause as StitchServiceException
-//                assertEquals(StitchServiceErrorCode.MONGODB_ERROR, svcEx.errorCode)
-//            }
-//
-//            try {
-//                Tasks.await(coll.findOneAndUpdate(Document(), Document("\$who", 1),
-//                        FindOneAndModifyOptions().upsert(true)))
-//                fail()
-//            } catch (ex: ExecutionException) {
-//                assertTrue(ex.cause is StitchServiceException)
-//                val svcEx = ex.cause as StitchServiceException
-//                assertEquals(StitchServiceErrorCode.MONGODB_ERROR, svcEx.errorCode)
-//            }
+            assertFails { findOneAndUpdate(Document(), Document("\$who", 1)).blockingGetResult() }
+            assertFails { findOneAndUpdate(Document(), Document("\$who", 1), FindOneAndModifyOptions().upsert(true)).blockingGetResult() }
         }
     }
 
     @Test
     fun find() {
         with(getCollectionInternal(COLLECTION_NAME)) {
-            // FIXME: fix find implementation - ignore this code for code review
-            val iter = find().blockingGetResult()!!
-            assertFalse(iter.iterator().hasNext())
-            assertFailsWith<NoSuchElementException> { iter.first() }
+            var iter = find()
+            assertFalse(iter.iterator().blockingGetResult()!!.hasNext().blockingGetResult()!!)
+            assertNull(iter.first().blockingGetResult())
+
+            val doc1 = Document("hello", "world")
+            val doc2 = Document("hello", "friend")
+            doc2["proj"] = "field"
+            insertMany(listOf(doc1, doc2)).blockingGetResult()
+
+            assertTrue(iter.iterator().blockingGetResult()!!
+                    .hasNext().blockingGetResult()!!)
+            assertEquals(doc1.withoutId(), iter.first().blockingGetResult()!!.withoutId())
+            assertEquals(doc2.withoutId(),
+                    iter.limit(1)
+                            .sort(Document("_id", -1))
+                            .iterator().blockingGetResult()!!
+                            .next().blockingGetResult()!!
```

Comment:
Does this _double asynchronicity_ make sense? Doesn't it only involve one remote call?

Reply:
That's right. I have changed it for the next commit - see https://github.com/realm/realm-java/pull/6881/files#r430916108